### ISSUE TITLE
Failed to reproduce WER when evaluating the downloaded wav2vec2 model

### ISSUE THREAD

**Issue by yasmin-z** [pytorch/fairseq#3658]
#### What is your question?
I tried to evaluate the model wav2vec_small_10m.pt (downloaded from the [wav2vec README page](https://github.com/pytorch/fairseq/tree/master/examples/wav2vec#readme)), using the 4-gram and transformer models for decoding, respectively, but the WERs I got was too much higher than results from the [paper](https://arxiv.org/abs/2006.11477).

A. When decoding with the 4-gram LM:
- e.g. my result: WER=96.2, evaluated on LibriSpeech test-clean dataset
          result from paper for same setting: WER=9.1

B. When decoding with the transformer LM:
- e.g. my result: WER=32.89, evaluated on dev-other dataset
          result from paper: WER=13.2

#### What have you tried?
I downloaded the wav2vec_small_10m.pt model (pre-trained on LS-960 dataset and fine-tuned on 10min librilight dateset), then followed instructions from the "[Evaluating a CTC model](https://github.com/pytorch/fairseq/tree/master/examples/wav2vec#evaluating-a-ctc-model)" section, and did the following:

Data preparation (evaluate with Librispeech test-clean for example):
1. generate .tsv file for test-clean dataset with: `python fairseq/examples/wav2vec/wav2vec_manifest.py /path/to/data/test-clean --dest /path/to/data --ext flac --valid-percent 0` (and rename to test-clean.tsv)
2. generate labels with: `python fairseq/examples/wav2vec/libri_labels.py /path/to/test-clean.tsv --output-dir /path/to/data --output-name test-clean`
3. download the [letter dictionary](https://dl.fbaipublicfiles.com/fairseq/wav2vec/dict.ltr.txt) provided

So, inside my 'data' directory, I have:
`test-clean  test-clean.tsv  test-clean.ltr  test-clean.wrd  dict.ltr.txt`

4. download & prepare LM
- (A) For decoding with 4-gram: download the 4-gram LM from the [wav2letter repository](https://github.com/flashlight/wav2letter/tree/master/recipes/sota/2019#pre-trained-language-models)
- (B) For decoding with transformer:
        1. download the transformer LM and transformer dictionary from the same [wav2letter repository](https://github.com/flashlight/wav2letter/tree/master/recipes/sota/2019#pre-trained-language-models)
        2. uppercase the transformer dictionary, name it 'dict.txt', and store it in the same directory as 'transformer.pt'
5. download the lexicon also from [wav2letter repository](https://github.com/flashlight/wav2letter/tree/master/recipes/sota/2019#tokens-and-lexicon-sets)

Then, for evaluation I ran:
- (A) using 4-gram:
```
python fairseq/examples/speech_recognition/infer.py /path/to/data --task audio_pretraining \
      --nbest 1 --path /path/to/wav2vec_small_10m.pt --gen-subset test-clean --results-path /path/to/results \
      --w2l-decoder kenlm --lm-model /path/to/4gram.bin --lm-weight 3.23 --word-score -0.26 --sil-weight 0 \
      --criterion ctc --labels ltr --max-tokens 4000000 --post-process letter \
      --lexicon /path/to/librispeech_lexicon.lst
```
This resulted in a WER of 96.2... (should be 9.1)

- (B) using transformer:
```
python fairseq/examples/speech_recognition/infer.py /path/to/data --task audio_pretraining \
      --nbest 1 --path /path/to/wav2vec_small_10m.pt --gen-subset test-clean --results-path /path/to/results \
      --w2l-decoder fairseqlm --lm-model /path/to/transformer.pt --lm-weight 1.2 --word-score -1.39 --sil-weight 0 \
      --criterion ctc --labels ltr --max-tokens 4000000 --post-process letter \
      --lexicon /path/to/librispeech_lexicon.lst
```
This resulted in WER of 25.8 (should be 6.9)

I copied the corresponding lm-weight and word-score directly from the wav2vec 2.0 paper.

No error was reported, and the hypo and ref files produced look fine, except that I can clearly see lots of word errors in the hypothesis. I'm not sure which part was wrong, could anyone please help me figure out why my WER is so high? 

Btw, I got same results with the paper when using viterbi decoding, so I guess there had been some problems with using the language model?

#### What's your environment?

 - fairseq Version: 1.0.0 
 - PyTorch Version: 1.9.0
 - OS: Ubuntu 18.04
 - How you installed fairseq: pip
 - Python version: 3.6.8
 - CUDA/cuDNN version: 11.1
 - installed flashlight python bindings from source

**Comment by homin-ryu** [comment]
Chapter 4.4 of the paper describes the beam. 
For using 4-gram, add `--beam 1500`.
For using transformer, add `--beam 500`.

**Comment by yasmin-z** [comment]
> Chapter 4.4 of the paper describes the beam.
> For using 4-gram, add `--beam 1500`.
> For using transformer, add `--beam 500`.

Thanks. The WER did improve to 61.5 after specifying beam to 1500 for 4-gram decoding.
Even so, it's still too high. Could there be any other parameters that I've missed?

ps. here's the last few lines of log output:
```
INFO:__main__:WER: 61.47101339013999
INFO:__main__:| Processed 2620 sentences (246320 tokens) in 895.4s (2.93sentences/s, 275.11 tokens/s)
INFO:__main__:| Generate test-clean with beam=1500
```

**Comment by yasmin-z** [comment]
Just discovered the reason being that the wrong 4-gram model was used. For 4-gram decoding, I previously downloaded the 4-gram model from the [wav2letter repository](https://github.com/flashlight/wav2letter/tree/master/recipes/sota/2019#pre-trained-language-models), but only the transformer model should be downloaded there. I had successfully reproduced WERs from the paper using the LibriSpeech official 4-gram model, downloaded from the [openslr site](https://openslr.magicdatatech.com/resources/11/) (and with beam=1500).
For decoding using the transformer model, increasing beam to 500 was the only change I made to get the WER right.

**Comment by Jxu-Thu** [comment]
@yasmin-z Could please provide the details for reproduction?
#3656 

> @Jxu-Thu please refer to the [#3658 (comment)](https://github.com/pytorch/fairseq/issues/3658#event-4957215645)


Can you provide the command to reproduce the results? And help check the url below is correct.
With kenlm, you download the LM and lexicon directly from the [https://openslr.magicdatatech.com/resources/11/](url)?
LM : [https://openslr.magicdatatech.com/resources/11/4-gram.arpa.gz](url)
Lexicon: [https://openslr.magicdatatech.com/resources/11/librispeech-lexicon.txt](url)
And how to set the lm-weight and word-score?

And How about Transformer LM?
1 LM : [https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_word_transformer.pt](url)
2 Lexicon: [https://dl.fbaipublicfiles.com/wav2letter/sota/2019/librispeech/models/am/librispeech-train%2Bdev-unigram-10000-nbest10.lexicon](url)
3 Download dict from [https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_word_transformer.dict](url). Upcase the transformer dictionary, name it 'dict.txt',

And how to set the lm-weight and word-score?

**Comment by yasmin-z** [comment]
> @yasmin-z Could please provide the details for reproduction?
> #3656
> 
> > @Jxu-Thu please refer to the [#3658 (comment)](https://github.com/pytorch/fairseq/issues/3658#event-4957215645)
> 
> Can you provide the command to reproduce the results? And help check the url below is correct.
> With kenlm, you download the LM and lexicon directly from the [https://openslr.magicdatatech.com/resources/11/](url)?
> LM : [https://openslr.magicdatatech.com/resources/11/4-gram.arpa.gz](url)
> Lexicon: [https://openslr.magicdatatech.com/resources/11/librispeech-lexicon.txt](url)
> And how to set the lm-weight and word-score?
> 
> And How about Transformer LM?
> 1 LM : [https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_word_transformer.pt](url)
> 2 Lexicon: [https://dl.fbaipublicfiles.com/wav2letter/sota/2019/librispeech/models/am/librispeech-train%2Bdev-unigram-10000-nbest10.lexicon](url)
> 3 Download dict from [https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_word_transformer.dict](url). Upcase the transformer dictionary, name it 'dict.txt',
> 
> And how to set the lm-weight and word-score?

Hi @Jxu-Thu,

I have mentioned the detailed steps to process data and run the inference [above](https://github.com/pytorch/fairseq/issues/3658#issue-931511956).
To be a bit more specific:

Download urls:
[Lexicon](https://dl.fbaipublicfiles.com/wav2letter/sota/2019/librispeech/models/am/librispeech-train%2Bdev-unigram-10000-nbest10.lexicon) and [Letter dictionary](https://dl.fbaipublicfiles.com/fairseq/wav2vec/dict.ltr.txt) for both;
1. 4-gram decoding: [LM](https://openslr.magicdatatech.com/resources/11/4-gram.arpa.gz)
2. transformer decoding: [LM](https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_word_transformer.pt) and [dictionary](https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_word_transformer.dict)

The lm-weight and word-score are from Appendix B of this [paper](https://arxiv.org/abs/2006.11477) (Table 7 for model pre-trained on LibriSpeech dataset).

Run from command line:
4-gram decoding:
```
python fairseq/examples/speech_recognition/infer.py /path/to/data --task audio_pretraining \
      --nbest 1 --path /path/to/model.pt --gen-subset $SUBSET --results-path /path/to/results \
      --w2l-decoder kenlm --lm-model /path/to/4gram.arpa.gz \
      --lm-weight $LM_WEIGHT --word-score $INSERTION_PENALTY --sil-weight 0 \
      --criterion ctc --labels ltr --max-tokens 4000000 --post-process letter --beam=1500 \
      --lexicon /path/to/librispeech_lexicon.lst
```
transformer decoding: 
```
python fairseq/examples/speech_recognition/infer.py /path/to/data --task audio_pretraining \
      --nbest 1 --path /path/to/model.pt --gen-subset $SUBSET --results-path /path/to/results \
      --w2l-decoder fairseqlm --lm-model /path/to/transformer.pt  \
      --lm-weight $LM_WEIGHT --word-score $INSERTION_PENALTY --sil-weight 0 \
      --criterion ctc --labels ltr --max-tokens 4000000 --post-process letter --beam=500 \
      --lexicon /path/to/librispeech_lexicon.lst
```

**Comment by arkhodamoradi** [comment]
This issue was very helpful for me to start with the tool and wav2vec. Either this post or something similar should be linked or added to the readme file for wav2vec.
